trace_decoder/src/decoding.rs

```diff
@@ -32,11 +34,120 @@ use crate::{
+/// inconsistencies found during processing, or any other condition that
+/// prevents successful completion of the trace processing task.
+#[derive(Debug)]
+pub struct TraceParsingError {
+    block_num: Option<U256>,
+    block_chain_id: Option<U256>,
+    txn_idx: Option<usize>,
+    addr: Option<Address>,
+    h_addr: Option<H256>,
+    slot: Option<U512>,
+    slot_value: Option<U512>,
+    reason: TraceParsingErrorReason, // The original error type
+}
+
+impl std::fmt::Display for TraceParsingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error processing trace: {}", self.reason)?;
+
+        if let Some(block_num) = self.block_num {
+            write!(f, "\nBlock num: {}", block_num)?;
+        }
+        if let Some(block_chain_id) = self.block_chain_id {
+            write!(f, "\nBlock chain id: {}", block_chain_id)?;
+        }
+        if let Some(txn_idx) = self.txn_idx {
+            write!(f, "\nTxn idx: {}", txn_idx)?;
+        }
+        if let Some(addr) = self.addr {
+            write!(f, "\nAddress: {}", addr)?;
+        }
+        if let Some(h_addr) = self.h_addr {
+            write!(f, "\nHashed address: {}", h_addr)?;
+        }
+        if let Some(slot) = self.slot {
+            write!(f, "\nSlot: {:x}", slot)?;
+        }
+        if let Some(slot_value) = self.slot_value {
+            write!(f, "\nSlot value: {:x}", slot_value)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl std::error::Error for TraceParsingError {}
+
+impl TraceParsingError {
+    /// Function to create a new TraceParsingError with mandatory fields
+    pub(crate) fn new(reason: TraceParsingErrorReason) -> Self {
+        Self {
+            block_num: None,
+            block_chain_id: None,
+            txn_idx: None,
+            addr: None,
+            h_addr: None,
+            slot: None,
+            slot_value: None,
+            reason,
+        }
+    }
+
+    /// Builder method to set block_num
+    pub(crate) fn block_num(&mut self, block_num: U256) -> &mut Self {
+        self.block_num = Some(block_num);
+        self
+    }
+
+    /// Builder method to set block_chain_id
+    pub(crate) fn block_chain_id(&mut self, block_chain_id: U256) -> &mut Self {
+        self.block_chain_id = Some(block_chain_id);
+        self
+    }
+
+    /// Builder method to set txn_idx
+    pub fn txn_idx(&mut self, txn_idx: usize) -> &mut Self {
+        self.txn_idx = Some(txn_idx);
+        self
+    }
+
+    /// Builder method to set addr
+    pub fn addr(&mut self, addr: Address) -> &mut Self {
+        self.addr = Some(addr);
+        self
+    }
+
+    /// Builder method to set h_addr
+    pub fn h_addr(&mut self, h_addr: H256) -> &mut Self {
+        self.h_addr = Some(h_addr);
+        self
+    }
+
+    /// Builder method to set slot
+    pub fn slot(&mut self, slot: U512) -> &mut Self {
+        self.slot = Some(slot);
+        self
+    }
```

Comment:
A bit related to my above comment, but it's pretty useful to also have the hashed slot. I wouldn't store it, but just display it along with the slot.

Reply:
do you mean I should hash it somehow?

Reply:
Yeah. You should just be able to call `hash(slot.as_bytes())` I think, but depending on the way we go with calculating the values internally or not, we can go either way.